ggml : WebGPU add TRANSPOSE and RESHAPE to supported ops

danbev opened the pull request:
This commit adds support for the TRANSPOSE and RESHAPE operations in the
ggml webgpu backend.

Co-authored-by: Diego Devesa <slarengh@gmail.com>

danbev committed:
ci : disable flash attention for webgpu test

This commit disables flash attention in the webgpu test.

The motivation for this is that it seem like flash attention might not
be supported for webgpu when using llvmpipe (not 100% though as it works
for me locally but I'm running a different version of mesa). This is an
snipped from the log:
```console
2025-08-31T10:49:20.2265119Z 27: /home/runner/work/llama.cpp/llama.cpp/ggml/src/ggml-backend.cpp:789: pre-allocated tensor (cache_v_l0 (view) (permuted) (transposed)) in a buffer (WebGPU) that cannot run the operation (TRANSPOSE)
2025-08-31T10:49:20.2266911Z 27: /home/runner/work/llama.cpp/llama.cpp/ggml/src/ggml-backend.cpp:789: pre-allocated tensor (cache_v_l0 (view) (permuted) (transposed)) in a buffer (WebGPU) that cannot run the operation (TRANSPOSE)
2025-08-31T10:49:20.2268797Z 27: ␛[34m0.01.085.256␛[0m ␛[35mW llama_context: layer 0 is assigned to device WebGPU but the Flash Attention tensor is assigned to device CPU (usually due to missing support)
2025-08-31T10:49:20.2269971Z 27: ␛[0m␛[34m0.01.085.262␛[0m ␛[35mW llama_context: Flash Attention was auto, set to disabled
2025-08-31T10:49:20.2271542Z 27: ␛[0m␛[34m0.01.085.302␛[0m ␛[35mW llama_context: layer 0 is assigned to device WebGPU but the Flash Attention tensor is assigned to device CPU (usually due to missing support)
2025-08-31T10:49:20.2272942Z 27: ␛[0m␛[34m0.01.085.303␛[0m ␛[35mW llama_context: Flash Attention was auto, set to disabled
2025-08-31T10:49:20.2274119Z 27: ␛[0m␛[34m0.01.085.334␛[0m ␛[35mW llama_context: layer 0 is assigned to device WebGPU but the Flash Attention tensor is assigned to device CPU (usually due to missing support)
2025-08-31T10:49:20.2275271Z 27: ␛[0m␛[34m0.01.085.335␛[0m ␛[35mW llama_context: Flash Attention was auto, set to disabled
2025-08-31T10:49:20.2276529Z 27: ␛[0m␛[34m0.01.085.470␛[0m ␛[35mW llama_context: layer 0 is assigned to device WebGPU but the Flash Attention tensor is assigned to device CPU (usually due to missing support)
2025-08-31T10:49:20.2277776Z 27: ␛[0m␛[34m0.01.085.471␛[0m ␛[35mW llama_context: Flash Attention was auto, set to disabled
2025-08-31T10:49:20.2279308Z 27: ␛[0m/home/runner/work/llama.cpp/llama.cpp/ggml/src/ggml-backend.cpp:789: pre-allocated tensor (cache_v_l0 (view) (permuted) (transposed)) in a buffer (WebGPU) that cannot run the operation (TRANSPOSE)
2025-08-31T10:49:20.2281488Z 27: /home/runner/work/llama.cpp/llama.cpp/ggml/src/ggml-backend.cpp:789: pre-allocated tensor (cache_v_l0 (view) (permuted) (transposed)) in a buffer (WebGPU) that cannot run the operation (TRANSPOSE)
```

danbev committed:
Revert "ci : disable flash attention for webgpu test"

This reverts commit 522ae983fe2c366eeb59f0096feae927d03d8621.

danbev committed:
tests : conditionally disable flash attention in webgpu

Just want to see if this enabled CI webgpu tests to pass.

slaren commented:
This is still a bug somewhere, it should not be hidden by disabling the test.

danbev committed:
Revert "tests : conditionally disable flash attention in webgpu"

This reverts commit 52794d367e17b7c724ad571a7fbc95f3a81cb780.

danbev committed:
ggml : WebGPU add TRANSPOSE and RESHAPE to supported ops

This commit adds support for the TRANSPOSE and RESHAPE operations in the
ggml webgpu backend.

Co-authored-by: Diego Devesa <slarengh@gmail.com>

CISC commented:
It should also be added here (and return `true` (not that it matters (yet) as the return is not checked)):
https://github.com/ggml-org/llama.cpp/blob/ea8412f854f9ed3d21a7db2636ceab42ee391b5c/ggml/src/ggml-webgpu/ggml-webgpu.cpp#L611-L614

danbev committed:
ggml : WebGPU added cases for TRANSPOSE and RESHAPE operations

This commit add GGML_OP_TRANSPOSE and GGML_OP_RESHAPE cases to the
ggml_webgpu_encode_node function in ggml-webgpu.cpp. The actual
operation are not implemented yet, and are left as TODOs.

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

danbev committed:
squash! ggml : WebGPU added cases for TRANSPOSE and RESHAPE operations [no ci]

Remove TODO comment about unimplemented operations.

danbev committed:
squash! ggml : WebGPU added cases for TRANSPOSE and RESHAPE operations [no ci]

Move GGML_OP_TRANSPOSE and GGML_OP_RESHAPE to the other no-op cases.
